File: server/common/protocol.py
import logging

STATUS_OK = 0x00
STATUS_ERROR = 0x01


MSG_TYPE_BATCH = 0x01
MSG_TYPE_QUERY_WINNERS = 0x02

def uint32_to_bytes_be(value):
    """Convert uint32 to 4 bytes big-endian"""
    return bytes([
        (value >> 24) & 0xFF,
        (value >> 16) & 0xFF,
        (value >> 8) & 0xFF,
        value & 0xFF
    ])

def uint16_to_bytes_be(value):
    """Convert uint16 to 2 bytes big-endian"""
    return bytes([
        (value >> 8) & 0xFF,
        value & 0xFF
    ])

def bytes_to_uint32_be(data):
    """Convert 4 bytes big-endian to uint32"""
    return (data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3]

def bytes_to_uint16_be(data):
    """Convert 2 bytes big-endian to uint16"""
    return (data[0] << 8) | data[1]


class BetRequest:
    def __init__(self, agencia_id, nombre, apellido, documento, nacimiento, numero):
        self.agencia_id = agencia_id
        self.nombre = nombre
        self.apellido = apellido
        self.documento = documento
        self.nacimiento = nacimiento
        self.numero = numero
    
    def __str__(self):
        return f"Bet({self.nombre} {self.apellido}, DOC:{self.documento}, NUM:{self.numero})"

class BatchRequest:
    def __init__(self, agencia_id, apuestas, is_last_batch=False):
        self.agencia_id = agencia_id
        self.apuestas = apuestas 
        self.is_last_batch = is_last_batch
    
    def __str__(self):
        return f"Batch(agency={self.agencia_id}, bets={len(self.apuestas)}, last={self.is_last_batch})"

class BetResponse:
    def __init__(self, status, message=""):
        self.status = status
        self.message = message
    
    def serialize(self):
        """Serialize response to binary format"""
        payload = bytearray()
        payload.append(self.status)
        
        message_bytes = self.message.encode('utf-8') if self.message else b''
        payload.append(len(message_bytes))
        payload.extend(message_bytes)
        
        length = len(payload)
        return uint32_to_bytes_be(length) + bytes(payload)

class QueryWinnersRequest:
    def __init__(self, agencia_id):
        self.agencia_id = agencia_id
    
    def __str__(self):
        return f"QueryWinners(agency={self.agencia_id})"

class WinnersResponse:
    def __init__(self, status, winner_dnis=None):
        self.status = status
        self.winner_dnis = winner_dnis or []
    
    def serialize(self):
        """Serialize winners response to binary format"""
        payload = bytearray()
        payload.append(self.status)
        
        winner_count = len(self.winner_dnis)
        payload.extend(uint16_to_bytes_be(winner_count))
        
        for dni in self.winner_dnis:
            payload.extend(uint32_to_bytes_be(dni))
        
        length = len(payload)
        return uint32_to_bytes_be(length) + bytes(payload)
    
    def __str__(self):
        return f"WinnersResponse(status={self.status}, winners={len(self.winner_dnis)})"

def recv_exactly(socket, length):
    """Receive exactly 'length' bytes from socket"""
    data = b''
    while len(data) < length:
        chunk = socket.recv(length - len(data))
        if not chunk:
            raise ConnectionError("Socket connection broken")
        data += chunk
    return data

def send_all(socket, data):
    """Send all data to socket"""
    total_sent = 0
    while total_sent < len(data):
        sent = socket.send(data[total_sent:])
        if sent == 0:
            raise ConnectionError("Socket connection broken")
        total_sent += sent

def read_string(data, offset):
    """Read length-prefixed string from data at offset (1 byte length)"""
    if offset >= len(data):
        raise ValueError("Unexpected end of data while reading string length")
    
    length = data[offset]
    if offset + 1 + length > len(data):
        raise ValueError("Unexpected end of data while reading string")
    
    string_data = data[offset + 1:offset + 1 + length]
    return string_data.decode('utf-8'), offset + 1 + length

def parse_bet_request(data):
    """Parse binary data into BetRequest object"""
    if len(data) < 4:
        raise ValueError("Message too short")
    
    msg_length = bytes_to_uint32_be(data[:4])
    
    if len(data) != 4 + msg_length:
        raise ValueError(f"Message length mismatch: expected {4 + msg_length}, got {len(data)}")
    
 
    payload = data[4:]
    offset = 0
    
    if offset + 1 > len(payload):
        raise ValueError("Unexpected end of data while reading agencia_id")
    agencia_id = payload[offset]
    offset += 1
    
    nombre, offset = read_string(payload, offset)
    
    apellido, offset = read_string(payload, offset)
    
    if offset + 4 > len(payload):
        raise ValueError("Unexpected end of data while reading documento")
    documento = bytes_to_uint32_be(payload[offset:offset + 4])
    offset += 4
    
    if offset + 4 > len(payload):
        raise ValueError("Unexpected end of data while reading nacimiento")
    anio = bytes_to_uint16_be(payload[offset:offset + 2])
    mes = payload[offset + 2]
    dia = payload[offset + 3]
    offset += 4
    nacimiento = anio * 10000 + mes * 100 + dia  # Reconstruct YYYYMMDD
    
   
    if offset + 2 > len(payload):
        raise ValueError("Unexpected end of data while reading numero")
    numero = bytes_to_uint16_be(payload[offset:offset + 2])
    
    return BetRequest(agencia_id, nombre, apellido, documento, nacimiento, numero)

def parse_batch_request(data):
    """Parse binary data into a BatchRequest object"""
    if len(data) < 4:
        raise ValueError("Message too short for batch")
    
    msg_length = bytes_to_uint32_be(data[:4])
    
    if len(data) != 4 + msg_length:
        raise ValueError(f"Message length mismatch: expected {4 + msg_length}, got {len(data)}")
    
    payload = data[4:]
    offset = 0
    
    if offset + 1 > len(payload):
        raise ValueError("Unexpected end of data while reading message type")
    msg_type = payload[offset]
    if msg_type != MSG_TYPE_BATCH:
        raise ValueError(f"Expected batch message type, got {msg_type}")
    offset += 1

    if offset + 1 > len(payload):
        raise ValueError("Unexpected end of data while reading agencia_id")
    agencia_id = payload[offset]
    offset += 1
    
    if offset + 1 > len(payload):
        raise ValueError("Unexpected end of data while reading is_last_batch")
    is_last_batch = payload[offset] == 1
    offset += 1
    
    if offset + 2 > len(payload):
        raise ValueError("Unexpected end of data while reading bet count")
    bet_count = bytes_to_uint16_be(payload[offset:offset + 2])
    offset += 2
    
    apuestas = []
    

    for i in range(bet_count):
        
        nombre, offset = read_string(payload, offset)
       
        apellido, offset = read_string(payload, offset)
        
        if offset + 4 > len(payload):
            raise ValueError(f"Unexpected end of data while reading documento for bet {i}")
        documento = bytes_to_uint32_be(payload[offset:offset + 4])
        offset += 4
        
        if offset + 4 > len(payload):
            raise ValueError(f"Unexpected end of data while reading nacimiento for bet {i}")
        anio = bytes_to_uint16_be(payload[offset:offset + 2])
        mes = payload[offset + 2]
        dia = payload[offset + 3]
        offset += 4
        nacimiento = anio * 10000 + mes * 100 + dia
        
        if offset + 2 > len(payload):
            raise ValueError(f"Unexpected end of data while reading numero for bet {i}")
        numero = bytes_to_uint16_be(payload[offset:offset + 2])
        offset += 2
        
        bet = BetRequest(agencia_id, nombre, apellido, documento, nacimiento, numero)
        apuestas.append(bet)
    
    return BatchRequest(agencia_id, apuestas, is_last_batch)

def receive_message(socket):
    """Receive and parse a complete message from socket"""
    try:
       
        length_bytes = recv_exactly(socket, 4)
        length = bytes_to_uint32_be(length_bytes)
        
        payload = recv_exactly(socket, length)
        
        complete_message = length_bytes + payload
        
        return parse_bet_request(complete_message)
            
    except (ValueError, ConnectionError) as e:
        logging.error(f"Error while receiving message: {e}")
        raise

def receive_batch_message(socket):
    """Receive and parse a batch message from socket"""
    try:
        length_bytes = recv_exactly(socket, 4)
        length = bytes_to_uint32_be(length_bytes)
        
        payload = recv_exactly(socket, length)
        
        complete_message = length_bytes + payload
        
        return parse_batch_request(complete_message)
            
    except (ValueError, ConnectionError) as e:
        logging.error(f"Error while receiving batch message: {e}")
        raise

def send_bet_response(socket, status, message=""):
    """Send bet response to client"""
    response = BetResponse(status, message)
    data = response.serialize()
    send_all(socket, data)

def parse_query_winners_request(data):
    """Parse binary data into a QueryWinnersRequest object"""
    if len(data) < 4:
        raise ValueError("Message too short for query winners request")
    
    msg_length = bytes_to_uint32_be(data[:4])
    
    if len(data) != 4 + msg_length:
        raise ValueError(f"Message length mismatch: expected {4 + msg_length}, got {len(data)}")
    
    payload = data[4:]
    offset = 0
    
    if offset + 1 > len(payload):
        raise ValueError("Unexpected end of data while reading message type")
    msg_type = payload[offset]
    if msg_type != MSG_TYPE_QUERY_WINNERS:
        raise ValueError(f"Expected query winners message type, got {msg_type}")
    offset += 1
    
    if offset + 1 > len(payload):
        raise ValueError("Unexpected end of data while reading agencia_id")
    
    agencia_id = payload[offset]
    
    return QueryWinnersRequest(agencia_id)

def receive_query_winners_message(socket):
    """Receive and parse a query winners message from socket"""
    try:
        length_bytes = recv_exactly(socket, 4)
        length = bytes_to_uint32_be(length_bytes)
        
        payload = recv_exactly(socket, length)
        
        complete_message = length_bytes + payload
        
        return parse_query_winners_request(complete_message)
            
    except (ValueError, ConnectionError) as e:
        logging.error(f"Error while receiving query winners message: {e}")
        raise

def send_winners_response(socket, status, winner_dnis=None):
    """Send winners response to client"""
    response = WinnersResponse(status, winner_dnis)
    data = response.serialize()
    send_all(socket, data)

File: client/common/protocol.go
package common

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
)

const (
	STATUS_OK    = 0x00
	STATUS_ERROR = 0x01
	
	MSG_TYPE_BATCH         = 0x01
	MSG_TYPE_QUERY_WINNERS = 0x02
)

type BetResponse struct {
	Status  uint8
	Message string
}

// QueryWinnersRequest represents a request to query winners for an agency
type QueryWinnersRequest struct {
	AgenciaID uint8
}

// WinnersResponse represents the response with winners for an agency
type WinnersResponse struct {
	Status       uint8
	WinnerCount  uint16
	WinnerDNIs   []uint32
}

// SerializeBatch serializes a batch of bets to binary
func SerializeBatch(batch *Batch) ([]byte, error) {
	payloadBuf := new(bytes.Buffer)
	

	binary.Write(payloadBuf, binary.BigEndian, uint8(MSG_TYPE_BATCH))
	
	binary.Write(payloadBuf, binary.BigEndian, batch.AgenciaID)
	
	var isLastByte uint8 = 0
	if batch.IsLastBatch {
		isLastByte = 1
	}
	binary.Write(payloadBuf, binary.BigEndian, isLastByte)
	
	binary.Write(payloadBuf, binary.BigEndian, uint16(len(batch.Apuestas)))

	for _, bet := range batch.Apuestas {
		if err := bet.Validate(); err != nil {
			return nil, fmt.Errorf("invalid bet in batch: %v", err)
		}
		
		if err := writeString(payloadBuf, bet.Nombre); err != nil {
			return nil, err
		}
		 
		if err := writeString(payloadBuf, bet.Apellido); err != nil {
			return nil, err
		}
		
		binary.Write(payloadBuf, binary.BigEndian, bet.Documento)
		
		binary.Write(payloadBuf, binary.BigEndian, uint16(bet.Nacimiento/10000)) 
		binary.Write(payloadBuf, binary.BigEndian, uint8((bet.Nacimiento/100)%100)) 
		binary.Write(payloadBuf, binary.BigEndian, uint8(bet.Nacimiento%100)) 
		
		binary.Write(payloadBuf, binary.BigEndian, uint16(bet.Numero))
	}
	
	payload := payloadBuf.Bytes()
	finalBuf := new(bytes.Buffer)
	binary.Write(finalBuf, binary.BigEndian, uint32(len(payload)))
	finalBuf.Write(payload)
	
	return finalBuf.Bytes(), nil
}

// SerializeBet serializes a bet to binary format with length prefix
func SerializeBet(bet *Bet) ([]byte, error) {
	if err := bet.Validate(); err != nil {
		return nil, err
	}
	
	payloadBuf := new(bytes.Buffer)
	
	if err := writeString(payloadBuf, bet.Nombre); err != nil {
		return nil, err
	}
	 
	if err := writeString(payloadBuf, bet.Apellido); err != nil {
		return nil, err
	}
	
	binary.Write(payloadBuf, binary.BigEndian, bet.Documento)
	
	binary.Write(payloadBuf, binary.BigEndian, uint16(bet.Nacimiento/10000)) 
	binary.Write(payloadBuf, binary.BigEndian, uint8((bet.Nacimiento/100)%100)) 
	binary.Write(payloadBuf, binary.BigEndian, uint8(bet.Nacimiento%100)) 
	
	binary.Write(payloadBuf, binary.BigEndian, uint16(bet.Numero))
	
	payload := payloadBuf.Bytes()
	finalBuf := new(bytes.Buffer)
	binary.Write(finalBuf, binary.BigEndian, uint32(len(payload)))
	finalBuf.Write(payload)
	
	return finalBuf.Bytes(), nil
}

func DeserializeBetResponse(data []byte) (*BetResponse, error) {
	if len(data) < 1 {
		return nil, fmt.Errorf("response too short")
	}
	
	response := &BetResponse{
		Status: data[0],
	}
	
	if len(data) > 1 {
		messageLen := int(data[1])
		if len(data) < 2+messageLen {
			return nil, fmt.Errorf("incomplete message")
		}
		response.Message = string(data[2 : 2+messageLen])
	}
	
	return response, nil
}

func writeString(buf *bytes.Buffer, s string) error {
	if len(s) > 255 {
		return fmt.Errorf("string too long: %d bytes (max 255)", len(s))
	}
	buf.WriteByte(byte(len(s)))
	buf.WriteString(s)
	return nil
}

// ReceiveResponse receives and deserializes a bet response from connection
func ReceiveResponse(reader io.Reader) (*BetResponse, error) {
	lengthData, err := recvAll(reader, 4)
	if err != nil {
		return nil, fmt.Errorf("failed to read response length: %v", err)
	}
	
	length := binary.BigEndian.Uint32(lengthData)
	
	responseData, err := recvAll(reader, int(length))
	if err != nil {
		return nil, fmt.Errorf("failed to read response payload: %v", err)
	}
	
	return DeserializeBetResponse(responseData)
}

// recvAll reads exactly 'length' bytes from connection
func recvAll(reader io.Reader, length int) ([]byte, error) {
	data := make([]byte, length)
	totalRead := 0
	
	for totalRead < length {
		n, err := reader.Read(data[totalRead:])
		if err != nil {
			return nil, err
		}
		if n == 0 {
			return nil, fmt.Errorf("connection closed")
		}
		totalRead += n
	}
	
	return data, nil
}

// SendBet serializes and sends a bet to connection
func SendBet(writer io.Writer, bet *Bet) error {
	data, err := SerializeBet(bet)
	if err != nil {
		return err
	}
	return sendAll(writer, data)
}

// sendAll writes all bytes to connection
func sendAll(writer io.Writer, data []byte) error {
	totalWritten := 0
	
	for totalWritten < len(data) {
		n, err := writer.Write(data[totalWritten:])
		if err != nil {
			return err
		}
		if n == 0 {
			return fmt.Errorf("connection closed")
		}
		totalWritten += n
	}
	
	return nil
}

// SendBatch sends a batch of bets to the server
func SendBatch(writer io.Writer, batch *Batch) error {
	data, err := SerializeBatch(batch)
	if err != nil {
		return fmt.Errorf("failed to serialize batch: %v", err)
	}
	
	return sendAll(writer, data)
}

// SerializeQueryWinners serializes a query winners request to binary
func SerializeQueryWinners(request *QueryWinnersRequest) ([]byte, error) {
	payloadBuf := new(bytes.Buffer)
	
	binary.Write(payloadBuf, binary.BigEndian, uint8(MSG_TYPE_QUERY_WINNERS))
	
	binary.Write(payloadBuf, binary.BigEndian, request.AgenciaID)
	
	payload := payloadBuf.Bytes()
	finalBuf := new(bytes.Buffer)
	binary.Write(finalBuf, binary.BigEndian, uint32(len(payload)))
	finalBuf.Write(payload)
	
	return finalBuf.Bytes(), nil
}

// DeserializeWinnersResponse deserializes a winners response from binary
func DeserializeWinnersResponse(data []byte) (*WinnersResponse, error) {
	if len(data) < 4 {
		return nil, fmt.Errorf("data too short for response length")
	}
	
	buf := bytes.NewReader(data)
	
	var length uint32
	if err := binary.Read(buf, binary.BigEndian, &length); err != nil {
		return nil, fmt.Errorf("failed to read response length: %v", err)
	}
	
	if len(data) != int(4+length) {
		return nil, fmt.Errorf("response length mismatch")
	}
	
	var status uint8
	if err := binary.Read(buf, binary.BigEndian, &status); err != nil {
		return nil, fmt.Errorf("failed to read status: %v", err)
	}
	
	var winnerCount uint16
	if err := binary.Read(buf, binary.BigEndian, &winnerCount); err != nil {
		return nil, fmt.Errorf("failed to read winner count: %v", err)
	}
	
	winnerDNIs := make([]uint32, winnerCount)
	for i := uint16(0); i < winnerCount; i++ {
		var dni uint32
		if err := binary.Read(buf, binary.BigEndian, &dni); err != nil {
			return nil, fmt.Errorf("failed to read DNI %d: %v", i, err)
		}
		winnerDNIs[i] = dni
	}
	
	return &WinnersResponse{
		Status:      status,
		WinnerCount: winnerCount,
		WinnerDNIs:  winnerDNIs,
	}, nil
}

// SendQueryWinners sends a query winners request
func SendQueryWinners(writer io.Writer, agenciaID uint8) error {
	request := &QueryWinnersRequest{AgenciaID: agenciaID}
	data, err := SerializeQueryWinners(request)
	if err != nil {
		return fmt.Errorf("failed to serialize query winners request: %v", err)
	}
	
	return sendAll(writer, data)
}

